### DISCUSSION TITLE
Collection is not available in CollectionBeforeChangeHook

### DISCUSSION THREAD

**Discussion by thgh** [payloadcms/payload#1380] [Roadmap]
# Bug Report

The typing of `CollectionBeforeChangeHook` suggests `collection` may be available, but it is not.
It is available in `CollectionBeforeOperationHook` -> `args.collection` so could be a quick win to pass it in `beforeChange` hooks

## Current Behavior

```ts
export const Pages = {
  slug: 'pages',
  hooks: {
    beforeChange: [({ req }) => {
      console.log('collection', req.collection?.config?.slug === 'pages') // false
    }]
  }
}
```

## Expected Behavior

```ts
export const Pages = {
  slug: 'pages',
  hooks: {
    beforeChange: [({ req }) => {
      console.log('collection', req.collection?.config?.slug === 'pages') // true
    }]
  }
}
```

## Possible Solution

Would be even handier if the collection object were passed at the toplevel instead of `req`

## Steps to Reproduce

1. See example

**Comment by jmikrut** [comment]
Hey @thgh — the types set `collection` as optional, as it is only available in REST API context and is added as middleware. I imagine you're using it in GraphQL?

I do believe we can inject the `collection` as an argument to all hooks, but for now, what we would typically do is make a higher-order function for your `beforeChange` hook that allows you to pass in the `slug` of your collection and then use that to retrieve your collection from `req.payload` (which is always there). 

```ts
const beforeChangeHook = (slug: string): CollectionBeforeChangeHook => ({ req: { payload }}) => {
  const collection = payload.collections[slug].config

  // do stuff here
}

const myCollectionSlug = 'my-collection'

const collection: CollectionConfig = {
  slug: myCollectionSlug,
  hooks: {
    beforeChange: [
      beforeChangeHook(myCollectionSlug),
    ]
  },
  // more collection stuff here
}
```

I'm going to convert this to a discussion and then mark it as planned so we can treat this as a feature request and get this built though.

👍 thanks for the idea!

**Comment by thgh** [comment]
I'm editing things from the admin UI which seems to be using the REST API. Cool to see it planned!
